File: ruby/lib/basecamp/services/cards_extensions.rb
# frozen_string_literal: true

module Basecamp
  module Services
    # Tri-state +due_on+ for card updates, prepended onto the generated
    # {CardsService} (see the +on_load+ hook in +basecamp.rb+).
    #
    # BC3's card controller is presence-aware on the JSON representation
    # (+kanban/cards_controller.rb+, basecamp/bc3#12521): +card_update_params+
    # is plain +card_params+, so an update writes exactly the keys the body
    # carries. An omitted +due_on+ leaves the card's due date UNCHANGED; an
    # explicit <tt>""</tt> (or +null+) clears it. The
    # <tt>{ due_on: nil }.merge(card_params)</tt> default survives only for the
    # HTML/turbo_stream web forms, which post every field on every submit.
    #
    # A clear therefore has to be *stated*, never encoded as an omission —
    # omitting +due_on+ to clear it is a silent no-op. <tt>""</tt> is the
    # spelling that travels: JSON +null+ cannot reach the wire from here at
    # all, because +compact_params+ is +kwargs.compact+ and drops nils (SPEC
    # section 18 body compaction). Rails casts the blank string to nil on the
    # date column, so <tt>""</tt> is what a clear looks like end to end.
    #
    # There is no read-before-write. An earlier version GET the card and resent
    # its due date, because the server then nil'd an unmentioned one and a
    # sparse PUT was destructive. Presence-awareness removed the hazard the
    # extra round-trip covered, and with it the race the round-trip opened
    # between the read and the write. Every case is a single PUT.
    module CardsExtensions
      # Updates a card, addressing only what the caller named.
      #
      # +due_on+ is tri-state:
      #
      # * +nil+ (unaddressed) — no +due_on+ key is sent; BC3 leaves the current
      #   due date alone
      # * <tt>""</tt> — a stated clear, sent as <tt>""</tt>
      # * a date — the due date is set
      #
      # Every other argument is plain send-when-set: +nil+ leaves the field off
      # the body, and BC3 leaves the stored value untouched.
      #
      # This is the same single PUT as {#update_verbatim}, which stays as the
      # unnormalised path; +update+ differs only in mapping an empty +due_on+ to
      # the <tt>""</tt> the server reads as a clear.
      #
      # @param card_id [Integer] card id
      # @param title [String, nil] new title (nil = keep current)
      # @param content [String, nil] new content (nil = keep current, "" clears)
      # @param due_on [String, nil] new due date (nil = keep current, "" clears)
      # @param assignee_ids [Array, nil] new assignees (nil = keep current, [] clears)
      # @return [Hash] the updated card
      def update(card_id:, title: nil, content: nil, due_on: nil, assignee_ids: nil)
        resolved_due_on =
          if due_on.nil?
            # Unaddressed. +compact_params+ drops the nil, so no key is sent and
            # the presence-aware update never touches the stored date.
            nil
          elsif due_on.to_s.empty?
            # A stated clear. "" survives +compact_params+ (it removes only
            # nils) and reaches the wire as {"due_on": ""}.
            ""
          else
            due_on
          end

        update_verbatim(
          card_id: card_id,
          title: title,
          content: content,
          due_on: resolved_due_on,
          assignee_ids: assignee_ids
        )
      end
    end
  end
end

File: typescript/src/services/cards-extensions.ts
import { CardsService as GeneratedCardsService } from "../generated/services/cards.js";
import type { Card } from "../generated/services/cards.js";

/**
 * Request parameters for `update`.
 *
 * Every field is presence-bearing: omitting it leaves that part of the card
 * alone. `dueOn` additionally accepts `null`, which is how you ask for the due
 * date to be *cleared* — a distinction the raw generated shape cannot express,
 * because its `dueOn` is a plain optional string and absence now means "leave
 * it alone".
 */
export interface UpdateCardRequest {
  /** Card title. Omit to leave unchanged. */
  title?: string;
  /** Card body (HTML). Omit to leave unchanged; `""` clears it. */
  content?: string;
  /**
   * Due date (YYYY-MM-DD).
   *
   * - omitted → the current due date is left unchanged
   * - `null` → the due date is cleared, sent on the wire as `"due_on": ""`
   * - a date → the due date is set
   */
  dueOn?: string | null;
  /**
   * Person IDs assigned to the card. Omit to leave the assignees alone; `[]`
   * clears them.
   *
   * Assignees are never resent on your behalf. BC3 filters incoming IDs
   * through `reachable_people`, so echoing back an ID belonging to someone who
   * has since lost board access would silently unassign them.
   */
  assigneeIds?: number[];
}

/**
 * CardsService with a presence-aware `update` on top of the generated surface
 * (`get`, `updateVerbatim`, `move`, ...).
 *
 * BC3's `kanban/cards_controller.rb` builds a JSON card update's params
 * straight from `card_params` (basecamp/bc3#12521), so on the JSON
 * representation an omitted key means **leave unchanged** — `due_on` included.
 * Clearing the due date is asked for explicitly: the server accepts JSON `null`
 * or `""`, and this SDK sends `""`, because a null would violate body
 * compaction (SPEC §18). Only the HTML/turbo_stream web forms still default an
 * omitted `due_on` to nil (`card_params.with_defaults(due_on: nil)`), which is
 * how the JSON path could be made presence-aware without disturbing the web
 * contract.
 *
 * `update` therefore sends exactly what the caller addressed, in a single PUT.
 * It differs from `updateVerbatim` only in accepting `dueOn: null` and encoding
 * it as the wire spelling of a clear.
 */
export class CardsService extends GeneratedCardsService {
  /**
   * Updates a card without disturbing fields you did not mention.
   *
   * One request. A field you omit is omitted from the body and the server
   * leaves it alone; `dueOn: null` clears the due date.
   *
   * Earlier releases read the card back first and resent its due date, because
   * a JSON update that omitted `due_on` used to erase it. bc3#12521 removed
   * that behaviour, so the extra round-trip — and the race between the read and
   * the write — is gone.
   *
   * @param cardId - The card ID
   * @param req - Fields to set; omitted fields are left unchanged
   * @returns The updated Card
   * @throws {BasecampError} If the request fails
   *
   * @example
   * ```ts
   * // Retitle without touching the due date.
   * await client.cards.update(123, { title: "New title" });
   *
   * // Clear the due date.
   * await client.cards.update(123, { dueOn: null });
   * ```
   */
  async update(cardId: number, req: UpdateCardRequest): Promise<Card> {
    const body: {
      title?: string;
      content?: string;
      dueOn?: string;
      assigneeIds?: number[];
    } = {};

    if (req.title !== undefined) body.title = req.title;
    if (req.content !== undefined) body.content = req.content;
    if (req.assigneeIds !== undefined) body.assigneeIds = req.assigneeIds;

    // `due_on` stays off the wire unless the caller addressed it, which the
    // server reads as "leave the due date alone". A date sets it; `null` asks
    // for a clear and is sent as `""` — the server casts the blank to nil (a
    // BC3 server test pins it), while `{"due_on": null}` would violate body
    // compaction (SPEC §18) and omission would now be silently ignored.
    if (req.dueOn !== undefined) body.dueOn = req.dueOn ?? "";

    return this.updateVerbatim(cardId, body);
  }
}
